### ISSUE TITLE
IPromiseBasedObservable typings improvement

### ISSUE THREAD

**Issue by Strate** [mobxjs/mobx-utils#41]
As for now, `IPromiseBasedObservable<T>` has this definition:
```ts
export interface IPromiseBasedObservable<T> {
    value: T;
    state: PromiseState;
    reason: any;
    promise: PromiseLike<T>;
    case<U>(handlers: {pending?: () => U, fulfilled?: (t: T) => U, rejected?: (e: any) => U}): U;
}
```

I suggest to change that to get stronger types:
```ts
interface IPromiseBasedObservablePending {
  state: "pending"
}
interface IPromiseBasedObservableFulfilled<T> {
  state: "fulfilled"
  value: T
}
interface IPromiseBasedObservableRejected {
  state: "rejected"
  value: Error
}
export type IPromiseBasedObservable<T> = (
  IPromiseBasedObservablePending |
  IPromiseBasedObservableFulfilled<T> |
  IPromiseBasedObservableRejected
) & {
  promise: PromiseLike<T>;
  case<U>(handlers: {pending?: () => U, fulfilled?: (t: T) => U, rejected?: (e: Error) => U}): U;
}
```

After that typescript will force developer to cover all cases of state and gives strict access to value:
```ts
function render(v: IPromiseBasedObservable<string>): string {
  return v.value // Property 'value' does not exists on type 'IPromiseBasedObservable<string>'
}

function render(v: IPromiseBasedObservable<string>): string {
    if (v.state === "pending") {
        return "value is pending"
    }
    if (v.state === "rejected") {
        return `value is rejected with: ${v.value.message}` // v.value is Error here
    }
    return `value is: ${v.value}`; // v.value is string here
}
```

Also I suggest to either:
1. Remove `case` function from IPromiseBasedObservable
2. Make all handlers required, for each state.

With strictNullChecks enabled it gives proper check for all cases of `v.state` are covered.
It is breaking change and usable only for TS2+, so, it should introduce new major version.

**Comment by jamiewinder** [comment]
I think the `value` type for `IPromiseBasedObservableRejected` should be `any` - you can throw / reject with anything in JavaScript after all. Looks good to me otherwise, though I still wonder about the decision to deprecate `reason`... not sure it makes sense to me.

**Comment by Strate** [comment]
@jamiewinder `reason` is not documented, so it would be removed.

**Comment by jamiewinder** [comment]
@Strate - yea, I know, I meant I don't understand why it was removed in the first place (nothing to do with your suggestion which seems sound to me)

**Comment by mweststrate** [comment]
Seems like a good improvement @Strate. Feel free to implement|!

**Comment by mweststrate** [comment]
Done and released as part of 3.0.0

**Comment by Strate** [comment]
Wow, great! Thank you, and sorry for not to be volunteer of this :(
